### PR TITLE
[SFT] add SFT Trainer Config dataclass

### PR DESCRIPTION
**Pull request by kashif**
Add a SFTConfig training argument dataclass together with deprecation warnings to the SFTTrainer

### CODE REVIEW

File: tests/test_sft_trainer.py [hunk 1]
```diff
@@ -1078,7 +1080,7 @@ def test_sft_trainer_llava(self):
...
+            processor.tokenizer.chat_template = """{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. {% for message in messages %}{% if message['role'] == 'user' %}USER: {% else %}ASSISTANT: {% endif %}{% for item in message['content'] %}{% if item['type'] == 'text' %}{{ item['text'] }}{% elif item['type'] == 'image' %}<image>{% endif %}{% endfor %}{% if message['role'] == 'user' %} {% else %}{{eos_token}}{% endif %}{% endfor %}{% if add_generation_prompt %}ASSISTANT: {% endif %}"""
```
Comment (lewtun) [inline tests/test_sft_trainer.py:1083]: Why is this being changed in this PR? It there a bug in the template we didn't see before?

File: trl/trainer/sft_config.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            Used only in case `dataset_text_field` is passed. This argument is used by the `ConstantLengthDataset` to pack the sequences
+            of the dataset. Defaults to False.
+    max_seq_length (`Optional[int]`):
+        The maximum sequence length to use for the `ConstantLengthDataset` and for automatically creating the Dataset. Defaults to min of the smaller of the `tokenizer.model_max_length` and `1024`.
+    dataset_num_proc (`Optional[int]`):
+            The number of workers to use to tokenize the data. Only used when `packing=False`. Defaults to None.
+    dataset_batch_size (`int`):
+        The number of examples to tokenize per batch. If batch_size <= 0 or batch_size == None,
+        tokenize the full dataset as a single batch. Defaults to 1000.
+    neftune_noise_alpha (`Optional[float]`):
```
Comment (lewtun) [inline trl/trainer/sft_config.py:36]: Can this be removed now that it's part of the base `TrainingArguments` @younesbelkada ?

File: trl/trainer/sft_trainer.py [hunk 3]
```diff
@@ -89,45 +90,33 @@ class SFTTrainer(Trainer):
...
-        dataset_kwargs: (`Optional[Dict]`, *optional*):
-            Dict of Optional kwargs to pass when creating packed or non-packed datasets
-        eval_packing: (`Optional[bool]`, *optional*):
-            Whether to pack the eval dataset as well. Defaults to `packing` if `None` is passed.
     """
 
     _tag_names = ["trl", "sft"]
 
+    @_deprecate_arguments(
+        version="0.10.0",
```
Comment (lewtun) [inline trl/trainer/sft_trainer.py:100]: Maybe set this to `1.0.0`?

**Review by lewtun — COMMENTED** [review]
Thanks a lot for adding this config @kashif ! LGTM apart from some nits. 

Given this is the most used trainer in the lib, would you mind adding a small backwards compatibility integration test where we init the trainer in two ways (setting all args we deprecated in the init) and then checking the results after 1 train step are the same?

File: trl/trainer/sft_config.py (hunk 2 above)
Comment (younesbelkada) [reply trl/trainer/sft_config.py:36]: Yeah totally ! we need a version check though on transformers so I think it's better to keep it ..

File: trl/trainer/sft_trainer.py [hunk 4]
```diff
@@ -24,6 +24,7 @@
 from datasets import Dataset
 from datasets.arrow_writer import SchemaInferenceError
 from datasets.builder import DatasetGenerationError
+from huggingface_hub.utils._deprecation import _deprecate_arguments
```
Comment (younesbelkada) [inline trl/trainer/sft_trainer.py:27]: Is it safe to use a private method ? Are we expecting any potential breaking change here?

**Review by younesbelkada — COMMENTED** [review]
Thank you so much ! In addition to @lewtun comments i have two tiny comments, WDYT?

File: tests/test_sft_trainer.py (hunk 1 above)
Comment (kashif) [reply tests/test_sft_trainer.py:1180]: yes i believe there was a bug in the chat template for llava... cc @edbeeching

Comment (edbeeching) [reply tests/test_sft_trainer.py:1180]: It was missing the gen prompt, not a "bug" per say, but a good idea to include it as people may copy-paste etc.

File: trl/commands/cli_utils.py [hunk 5]
```diff
@@ -142,11 +142,6 @@ def warning_handler(message, category, filename, lineno, file=None, line=None):
 @dataclass
 class SftScriptArguments:
     dataset_name: str = field(default="timdettmers/openassistant-guanaco", metadata={"help": "the dataset name"})
-    dataset_text_field: str = field(default=None, metadata={"help": "the text field of the dataset"})
```
Comment (younesbelkada) [inline trl/commands/cli_utils.py:145]: why this has been removed? 🤔

Comment (kashif) [reply trl/commands/cli_utils.py:145]: let me fix it...

Comment (younesbelkada) [reply trl/commands/cli_utils.py:145]: ok ok ! thanks !

**Review by younesbelkada — APPROVED** [review]
Looking great thanks ! failing test is unrelated to this PR: https://github.com/huggingface/trl/actions/runs/8786654281/job/24110021581?pr=1530

File: trl/commands/cli_utils.py [hunk 6]
```diff
@@ -142,11 +142,8 @@ def warning_handler(message, category, filename, lineno, file=None, line=None):
...
     dataset_train_name: str = field(default="train", metadata={"help": "the name of the training set of the dataset"})
     dataset_test_name: str = field(default="test", metadata={"help": "the name of the training set of the dataset"})
```
Comment (lewtun) [inline trl/commands/cli_utils.py:146]:
```suggestion
    dataset_test_split: str = field(default="test", metadata={"help": "The dataset split to evaluate on"})
```

**Review by lewtun — APPROVED** [review]
Looks great, with a minor nit on the dataset split names. Feel free to merge after fixing that!

**Pull request merged**